src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs

```diff
@@ -222,6 +224,65 @@ public static UIntPtr Exchange(ref UIntPtr location1, UIntPtr value)
             return (UIntPtr)Exchange(ref Unsafe.As<UIntPtr, int>(ref location1), (int)value);
 #endif
         }
+
+        /// <summary>Sets a variable of the specified type <typeparamref name="T"/> to a specified value and returns the original value, as an atomic operation.</summary>
+        /// <param name="location1">The variable to set to the specified value.</param>
+        /// <param name="value">The value to which the <paramref name="location1"/> parameter is set.</param>
+        /// <returns>The original value of <paramref name="location1"/>.</returns>
+        /// <exception cref="NullReferenceException">The address of location1 is a null pointer.</exception>
+        /// <exception cref="NotSupportedException">An unsupported <typeparamref name="T"/> is specified.</exception>
+        /// <typeparam name="T">
+        /// The type to be used for <paramref name="location1"/> and <paramref name="value"/>.
+        /// This type must be a reference type, an enum type (i.e. typeof(T).IsEnum is true), or a primitive type (i.e. typeof(T).IsPrimitive is true).
+        /// </typeparam>
+        [return: NotNullIfNotNull(nameof(location1))]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static T Exchange<T>([NotNullIfNotNull(nameof(value))] ref T location1, T value)
+        {
+            // Handle all reference types with CompareExchange(ref object, ...).
+            if (!typeof(T).IsValueType)
+            {
+                object? result = Exchange(ref Unsafe.As<T, object?>(ref location1), value);
+                return Unsafe.As<object?, T>(ref result);
+            }
+
+            // Handle everything else with a CompareExchange overload for the unsigned integral type of the corresponding size.
+            // Only primitive types and enum types (which are backed by primitive types) are supported.
+            if (!typeof(T).IsPrimitive && !typeof(T).IsEnum)
+            {
+                throw new NotSupportedException(SR.NotSupported_ReferenceEnumOrPrimitiveTypeRequired);
+            }
+
+            if (Unsafe.SizeOf<T>() == 1)
```

Comment:
This can be `sizeof(T)`. It is cheaper than `Unsafe.SizeOf<T>`.

Reply:
Should codegen lower one into the other (where possible)?

Reply:
The final code should be the same. `sizeof` is cheaper for startup. `sizeof` is IL opcode. ``Unsafe.SizeOf` is JIT intrinsic. It requires importing the call of the method, creating the generic method instantiation and recognizing the method by name as intrinsic, ... . None of this is necessary for the `sizeof` opcode.